### DISCUSSION TITLE
Android Build Method

### DISCUSSION THREAD

**Discussion by marchbold** [airsdk/Adobe-Runtime-Support#1782] [General]
I just want to put this here as information for anyone using our latest extensions. 

We have updated all our extensions to use the latest Android features. We continue to have issues with resources using the legacy build method and have decided that the effort to maintain this too high so we are only going to support the new build method going forward. Anyone using the legacy build method will need to continue using the older version of our extensions, (prior to February 2022). 

For anyone who is unaware, Harman has implemented a completely new build system using gradle and the Android SDK for building AAB and has extended this to the APK packaging. You can still specify to use the legacy build method if for some reason you require it.  This new approach is how most other platforms (unity, flutter etc) create Android applications so it brings AIR access to a much broader feature set in the future.

The new build method should be the default, however to ensure you are using the new build method:

1. You must remove the `<BuildLegacyAPK>true</BuildLegacyAPK>` flag from your app descriptor if you have added it.

2. With the latest release of AIR (33.1.1.795) it appears the default method has gone back to using the legacy implementation, which will result in resource errors when building an apk like the below:

```
xxxxx\lifecycle-viewmodel-2.4.1-res\values\values.xml:3: error: Found tag id where item is expected
xxxxx\lifecycle-runtime-2.4.1-res\values\values.xml:3: error: Found tag id where item is expected
...

```

@ajwfrost has posted the following update to correct this behaviour: https://github.com/airsdk/Adobe-Runtime-Support/files/8230970/adt.zip (from [here](https://github.com/airsdk/Adobe-Runtime-Support/discussions/1762#discussioncomment-2338157))

**Comment by ajwfrost** [comment]
To clarify the behaviour/logic on whether we go with the 'legacy' build method, or use the Gradle one for APKs:
- originally we would go with the Gradle method unless opting out via `<BuildLegacyAPK>`, but there were a lot of issues/complaints!
- we then added code to check that people had the platform SDK (i.e. Android SDK and build tools) and appropriate Java version set up before we go with Gradle (and also added code to check for these in their default installation locations)
- finally though, we added a check in 33.1.1.795 on the application descriptor version, and if that's less than 33.1 then the legacy method will be used. This is where the bug mentioned above was introduced, which means folk who took the 33.1.1.795 build within a day or so of its release will always be getting the legacy mode unless they download the patch; the current 33.1.1.795 zip files now include that patch

We may of course have gone too far in trying to ensure a smooth transition (or rather, a lack of transition) for people who are just rebuilding their existing applications with a later AIR SDK, particularly given the changes in ANE and resource handling that were associated with this. So basically: the best thing to do is to update your application descriptor files to 33.1 and ensure it's using the Gradle build mechanism (if you use legacy mode, you'll probably be seeing a warning about captive runtime packaging) - for which you need to install Android Studio and the Android SDK.



Another thing to note - the default Android target SDK version was set to 31 (i.e. Android 12) when we updated our own code to work with this, but this is triggering some changes to Android behaviour so it may be worth setting this back to 30 via your application descriptor files (e.g. `<uses-sdk android:targetSdkVersion="30" />`)

thanks

**Reply by marchbold** [reply]
@ajwfrost Out of interest, what issues/complaints were there with the new build method?

**Reply by ajwfrost** [reply]
Mostly it was trying to get set up with appropriate values for Java and for the platform SDK, a lot of confusions because of "SDK" now meaning either the AIR SDK or the Android SDK, plus different versions of the Android SDK seem to have different paths for ADB etc. So mostly it was just a sense of having to do a lot of extra set-up and configuration in order to get it working, when previously it had just all been working fine with the older mechanism...

**Reply by marchbold** [reply]
Ah okay, that makes sense. Hopefully we can iron out the required android sdk installation process as far as I can tell this is the way all the other cross-platform tech works and there are a lot of benefits to this approach.

**Reply by ajwfrost** [reply]
Yes, it would be great if we can support people through the steps here :-)

As a slight aside ... do you see a benefit in something similar for macOS and/or iOS (on mac) development? We made some changes a while ago with how the start-up works on mac, and with some of the recent issues folk have had, I wonder whether it's better to create an Xcode project and let people then build/sign/deploy that way. It's nice to do everything from IntelliJ, but it's not really working now that Apple have extra requirements on signing etc...

thanks!

**Reply by marchbold** [reply]
On the fence with that one, I do see some advantages but it adds a massive layer of complexity. Unity takes this approach but we often run into issues with version conflicts and then post build tools that don't work correctly. We have spent more time working on post build scripts to handle this environment in Unity than any other area of our extensions.

Also we have found with Unity that this approach extends a build time significantly. Much more than the android implementation did. To the point where I'd actually test code in an AIR extension before moving it into Unity to test the final integration. AIR builds I would guess are 5-10 times faster from code change to a test on device. 

This one would have to involve a large beta program and extensive testing.

**Reply by ajwfrost** [reply]
Thanks for the insights! Good to know :-)

**Reply by saykow** [reply]
is there any tutorial how to set up this?

**Reply by ajwfrost** [reply]
"this" being the Android build system? Do you already have the latest AIR SDK and tools installed? If so, then to use the new build system, you need:
1. an Android SDK - ideally installed in the default location for your platform, or with the location mentioned in the AIR SDK configuration file or command line
2. to be using Java Development Kit 11 or later (so not just a JRE or an earlier JDK) - if you have Android Studio installed in the default location, it should come with JDK11 and we'd pick up this path, or you can specify the location in the AIR SDK configuration file or via a JAVA_HOME environment variable
3. to have your AIR application descriptor XML file set up with the "33.1" namespace and to not include the `BuildLegacyAPK` instruction

If you have any issues, please let us know what you see in the output, or preferably if you can enable the logging in ADT via the configuration file, we can see what's happening internally..

thanks

**Comment by saykow** [comment]
Hello! I have a web flash game and I wanted to put it on android. Im trynna
make it like a mobile launcher for that game. But I dont know how to setup
or configure everything. Thats why im looking for a step by step guide to
put my game and build this application for android.

**Reply by ajwfrost** [reply]
Ah - that might be worth a new discussion of its own... there are a few folk that have done that, I saw that you can still play the Flash version of Doom which has been wrapped into an AIR application now...

There are likely to be a number of "gotcha" issues with things like loading resources and permissions... so just trying to load a SWF that's on a webserver and then have this download resources may run into problems, but packaging the whole app as an AIR application should work (of course, any online/server component would need revising).

Do you have this game online still?

**Comment by saykow** [comment]
Yes I do. Can you help me sir make a launcher for my game? Badly need it.

**Reply by ajwfrost** [reply]
Can you please email me? andrew.frost at harman.com - if you can email me the link for the game I can have a quick look at what may be the best approach for this.

**Comment by saykow** [comment]
***@***.*** sir? Is that correct? Okay

**Comment by httpwebmedia** [comment]
What if we want to use APM with legacy method? What do we need to set?

**Reply by marchbold** [reply]
apm doesn't change or add the BuildLegacyAPK flag. You need to add it manually if you wish to use it.

Note: you won't be able to use our latest ANEs with the legacy build method.

**Reply by httpwebmedia** [reply]
Yes I read it in the doc but I managed to convert the old 1024 certificates and port to play app signing (finally).
This way I'll get rid of old Java 1.8 and hopefully maybe the ANR issues are related to the use of old java to compile the apps?
Would that be a possibility of all these ANR issues?

**Reply by marchbold** [reply]
I don't think they would be related to the java version directly, but upgrading java will allow you to use the latest build method and AIR SDK so you should see improvements there at least. 

My experience from a lot of these ANRs' is that it can be worked around by changing the way you do things in your app. Image loading, and audio playback process are the biggest ones. Also ANE loading at start hopefully has some fixes in the works from Harman.

**Comment by xxxVBxxx** [comment]
Hello, 
I set up everything as you said above, but when I'm typing the android keystore password, it throws an exception saying "incorrect password".
It turns out that the keystore file was created using jdk 1.8 and I asume jdk11 encrypts the password differently. 
Is there anyway to fix this issue without a new keystore file?

**Reply by marchbold** [reply]
Should be no problem reading a certificate in 11 created by 8, as long as it's not 1024, they dropped support for that in java > 1.8_112

**Reply by xxxVBxxx** [reply]
Thanks for replying @marchbold 
Is there a workaround for this issue if it's 1024?

**Reply by marchbold** [reply]
There's no real workaround, you have to move to a new certificate for use with java 11. Regarding handling issues with Google Play applications you can either:
- move to play app signing (upload your 1024 cert to google and create a new upload certificate)
- create a temp cert to use and then resign your package using java 1.8.112 after packaging. 

We use the first method and have had good success with it. Biggest issue is that you have to add the upload certificate details to any auth services that verify the signature (eg facebook or google sign in etc) so they work while debugging.

**Reply by xxxVBxxx** [reply]
I created and uploaded the encrypted_certificate.zip file and now can't figure out what to do next. I need to request key upgrade in order to be able to upload the upload certificate file, right? 
Sorry for a dumb question but I'm just not sure what is the correct way to do and don't want to mess it up.
Thanks in advance
![изображение](https://user-images.githubusercontent.com/40067995/179519052-23b4b786-b917-4d45-ac6e-ef65ff4f05ec.png)

![изображение](https://user-images.githubusercontent.com/40067995/179518016-ce164f14-b2fe-4f4a-835e-4e461a002491.png)

**Reply by marchbold** [reply]
If you are upgrading to play app signing you can do the upload and app signing keys at the same time. You should see an option just after you upload your app signing key, something like "For increased security , create a new upload key ... " and an option to expand it.

If you don't they will use your app signing key as the upload key and you'll need to request an upgrade if you wish to upload a separate one. 

Do not request an "app signing key" upgrade, this has larger consequences. 

More details here: https://airsdk.dev/docs/tutorials/platform/android/packaging-android-app-bundles#signing-and-certificates

**Reply by xxxVBxxx** [reply]
Unfortunately we have already done the "Export and upload a key from Java keystore" and uploaded the  encrypted_certificate.zip file but didn't do the "For increased security , create a new upload key" part and I can't find a way to do now.
Now the app signing page looks like this. No option for uploading the upload certificate file.
![изображение](https://user-images.githubusercontent.com/40067995/179686344-ccefdbd8-e31d-4477-a7cf-aefc672787ae.png)

**Reply by marchbold** [reply]
Sure, you'll have to go through the request process then, from memory it's confusing as it seems like you are resetting the app signing one at the start but then there's an option to do the upload cert on the next step. 

Was a while since I've done my last one though..

**Reply by xxxVBxxx** [reply]
That's what I'm afraid of. When I select "Request key upgrade" and choose "Upload a new app signing key from Java keystore" there is no "create a new upload key" option.  

![изображение](https://user-images.githubusercontent.com/40067995/179691727-9acc53fa-ae0e-470b-bb6c-96e64d99333e.png)

**Reply by marchbold** [reply]
Right , yeah that really looks like the app signing key not the upload key.

**Reply by marchbold** [reply]
So looks like you are clicking the "Request key upgrade" in the app signing key section? 

You need to go further down, in the Upload key certificate section and click the "contact our support team" 

![Screen Shot 2022-07-19 at 9 43 59 pm](https://user-images.githubusercontent.com/442356/179742435-68e1fb62-27cf-41f7-affe-6ae841f8aa28.png)

It will give you a form you have to fill out and they will send you a reset link .

**Reply by xxxVBxxx** [reply]
@marchbold thanks a lot for your help: I finnaly managed to build and upload the android app.
Do I need to build both armv7 and armv8 for 32bit and 64bit devices or only one build for all devices?

**Reply by marchbold** [reply]
You should be uploading an aab now which contains both, and likely x86/x64 as well.

**Comment by mgravano** [comment]
Hi there,
I am using Animate for my application and the latest version of Air SDK 50.0.0
Could you please tell me, if possible, how I implement the new build method in Animate. Currently I simply import the SDK in the Help/Manage SDK section of Adobe AIR. Is there any other way to import the SDK intoAnimate?

**Reply by mgravano** [reply]
I already solved it, thanks

**Reply by franmocad** [reply]
Could you please tell me how

**Reply by mgravano** [reply]
I thought I had solved it but it is not.
I installed Android Studio on my PC and included an environment variable AIR_ANDROID_SDK_HOME which pointed to Android Studio's Android SDK.
At that time I tried to publish my app and the apk file was created without errors, so I thought I had solved the problem.
But then I tried to open the apk file on my phone and it fails.
I will keep trying, any guide to publish my app will be welcome.
Thank you!

**Reply by ajwfrost** [reply]
The build method we're talking about here is for packaging up the application after it's been compiled. So from Animate's perspective, it creates the SWF file and then calls ADT in order to package it - the packaging is where we're now using the Android tools for this.

There are two things needed here, (a) an Android SDK, and (b) a Java development kit. Both of these should be present on your machine if you've got Android Studio and done the normal set-up steps to get an SDK.

The other thing that we check is the app descriptor version so please ensure you're using namespace "50.0" for the app descriptor file.

If it's still failing, please try enabling the debug output (via the AIR SDK config file under your home folder's ".airsdk" locaiton - e.g c:\users\username\.airsdk\adt.cfg) - remove the comment from before the `DebugOut=true` - and then run it again. There should be some output file then at c:\users\username\adt.log that tells you what's going on, and if it's not using the Gradle build approach it will tell you why...

thanks

**Comment by mgravano** [comment]
Hi,
Thanks for your answer!
I have the latest version of Android Studio and the Android Studio Java Development kit installed. I created environment variables pointing to these:

![environment-variables](https://user-images.githubusercontent.com/5115296/201696586-c0b2f102-9884-4bc8-8a9c-fca7bd3e0145.jpg)

Animate version is also latest and uses JAVA_VERSION="11.0.15.1"

I am using namespace "50.0" for the app descriptor file:
`<application xmlns="http://ns.adobe.com/air/application/50.0">`

I removed the comment before DebugOut=true from the adt.cfg file before publishing.

But I still can't open the apk file on my phone...

This is the adt.log

[adt.log](https://github.com/airsdk/Adobe-Runtime-Support/files/10004344/adt.log)

Thanks for your help!

**Reply by ajwfrost** [reply]
Well everything is set up in terms of ADT etc, and that is using the new mechanism:
```
Gradle build command:  C:/Program Files/Adobe/Adobe Animate 2023/jre\bin\java.exe -Xmx1024m -Dorg.gradle.appname=gradlew -Dorg.gradle.jvmargs=-Xmx1024m -Duser.dir="C:\Users\User\AppData\Local\Temp\cd77fc3a-ba0d-41ba-8c08-214645b03697" -cp "C:\Users\User\AppData\Local\Temp\cd77fc3a-ba0d-41ba-8c08-214645b03697\gradle\wrapper\gradle-wrapper.jar";"C:\Users\User\Desktop\clientes\EL BUHO BOO\App nueva\AIRSDK_Windows\lib\android\bin\gradle-cli.jar" org.gradle.wrapper.GradleWrapperMain :app:assembleRelease
```

Slightly concerned about your setting for the "extdir" value, are you able to ensure Animate is pointing properly to the folders where your latest .ane files are?

Otherwise .. it looks the deployment and installation works fine. If the app doesn't start up properly, then you need to start debugging on the phone. I'd suggest first using `apk-debug` for the build type, to give more info, and then once you've installed it onto the device you can do `adt -deviceLog -platform android` to start capturing traces from it, and then launch it by clicking on the icon on the device (or usd ADB/Android Studio directly if you prefer).

thanks

**Reply by mgravano** [reply]
I have my anes in C:\Users\User\ane

But there are also anes in C:\Users\User\apm_packages

I used the ones in the C:\Users\User\ane folder, is that ok? Which ones should I use?

Thanks

**Reply by mgravano** [reply]
I tried with the ane in both locations and the result is the same. I'm going to try to capture traces by starting the phone.

**Reply by mgravano** [reply]
I already captured the traces by starting the app on the phone.

[logs.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/10006208/logs.txt)

**Reply by marchbold** [reply]
You should only add the ones in the `ane` directory from apm. The `apm_packages` dir is a cache for your air packages and shouldn't be used directly

**Reply by mgravano** [reply]
Thanks!

**Reply by marchbold** [reply]
@mgravano I've responded in the Adverts forum as well, but this crash is a forced crash caused by the admob sdk when you provide an invalid admob app id. Follow the steps in the Adverts forum to correct it

**Comment by mgravano** [comment]
Thanks

**Comment by applauzmedia** [comment]
Does anyone have a step by step guide as for what to do on a Mac with Adobe Animate to publish for Android. I've read this thread but I'm still at a loss as to what exactly I need to do.  I'm using AIR SDK 50.

Any help is appreciated.

**Comment by jpgrovy** [comment]
I am now able to get the app to build to an Android phone using the Android Studio solution ... but the app crashes immediately when opening. Crashes too soon to get a trace. There is no code in my app yet.... only the appropriate import statements.

**Reply by marchbold** [reply]
Check the device logs, will likely be an issue with the manifest such as an activity that doesn't exist.

https://docs.airnativeextensions.com/docs/tutorials/device-logs

**Comment by truth2022-1A** [comment]
No, you can't! The short answer is NO!
